pkg/virt-controller/network/ipamclaims.go

```diff
@@ -0,0 +1,195 @@
+	if err != nil {
+		return fmt.Errorf("failed composing networkToIPAMClaimName: %w", err)
+	}
+
+	// For a standalone VMI, the IPAMClaim owner is the VMI itself
+	if ownerRef == nil {
+		ownerRef = v1.NewControllerRef(vmi, virtv1.VirtualMachineInstanceGroupVersionKind)
+	}
+
+	claims := composeIPAMClaims(vmi.Namespace, ownerRef, networkToIPAMClaimParams)
+	err = createIPAMClaims(client, vmi.Namespace, claims)
+	if err != nil {
+		return fmt.Errorf("failed IPAMClaims creation for VMI %s: %w", vmi.Name, err)
+	}
+
+	return nil
+}
+
+func composeIPAMClaims(namespace string, ownerRef *v1.OwnerReference, networkToIPAMClaimParams map[string]IPAMClaimParams) []*ipamclaims.IPAMClaim {
+	claims := []*ipamclaims.IPAMClaim{}
+	for netName, ipamClaimParams := range networkToIPAMClaimParams {
+		claims = append(claims, composeIPAMClaim(
+			namespace,
+			*ownerRef,
+			ipamClaimParams,
+			namescheme.GenerateHashedInterfaceName(netName),
+		))
+	}
+
+	return claims
+}
+
+func createIPAMClaims(client kubecli.KubevirtClient, namespace string, claims []*ipamclaims.IPAMClaim) error {
+	for _, claim := range claims {
+		_, err := client.IPAMClaimsClient().K8sV1alpha1().IPAMClaims(namespace).Create(
+			context.Background(),
+			claim,
+			v1.CreateOptions{},
+		)
+
+		if err != nil {
+			if !k8serrors.IsAlreadyExists(err) {
+				return fmt.Errorf("failed to create IPAMClaim: %w", err)
+			}
+
+			err = validateIPAMClaim(client, namespace, claim.Name, claim.OwnerReferences[0].UID)
+			if err != nil {
+				return fmt.Errorf("failed validating IPAMClaim: %w", err)
+			}
+		}
+	}
+
+	return nil
+}
+
+func composeIPAMClaim(namespace string, ownerRef v1.OwnerReference, ipamClaimParams IPAMClaimParams, interfaceName string) *ipamclaims.IPAMClaim {
+	return &ipamclaims.IPAMClaim{
+		ObjectMeta: v1.ObjectMeta{
+			Name:      ipamClaimParams.claimName,
+			Namespace: namespace,
+			OwnerReferences: []v1.OwnerReference{
+				ownerRef,
+			},
+		},
+		Spec: ipamclaims.IPAMClaimSpec{
+			Network:   ipamClaimParams.networkName,
+			Interface: interfaceName,
+		},
+	}
+}
+
+func filterNonAbsentNetworks(interfaces []virtv1.Interface, networks []virtv1.Network) []virtv1.Network {
+	nonAbsentIfaces := vmispec.FilterInterfacesSpec(interfaces, func(iface virtv1.Interface) bool {
+		return iface.State != virtv1.InterfaceStateAbsent
+	})
+	nonAbsentNetworks := vmispec.FilterNetworksByInterfaces(networks, nonAbsentIfaces)
+
+	return nonAbsentNetworks
+}
+
+func validateIPAMClaim(client kubecli.KubevirtClient, namespace string, claimName string, expectedOwnerUID types.UID) error {
+	currentClaim, err := client.IPAMClaimsClient().K8sV1alpha1().IPAMClaims(namespace).Get(
+		context.Background(),
+		claimName,
+		v1.GetOptions{},
+	)
+	if err != nil {
+		return fmt.Errorf("failed getting IPAMClaim: %w", err)
+	}
+
+	// We need to make sure that the existing IPAMClaim belong to current VMI,
+	// and it isn't a lefotver of freshly deleted VM/VMI with the same namespace/name
```

Comment:
maybe somehow embed this information in the function's name ? For clarity: I propose renaming this func so its purpose is immediately clear for the reader when checking the client code.

Let's at least try :)